### PR TITLE
Blockstore api

### PR DESCRIPTION
**Pull request by monsieurleberre**
First REST function for the blockstore

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.BlockStore/Models/BlockByHashQueryModel.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Text;
+
+namespace Stratis.Bitcoin.Features.BlockStore.Models
+{
+    public abstract class QueryModelBase
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.BlockStore/Models/BlockByHashQueryModel.cs]: For consistency, look at how we created and named those in https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Wallet/Models/RequestModels.cs

File: src/NBitcoin/Block.cs [hunk 2]
```diff
@@ -428,6 +429,23 @@ public static Block ParseJson(string json)
             return blk;
         }
 
+        public string ToJson(bool indented = false)
```
Comment (bokobza) [inline src/NBitcoin/Block.cs]: I actually really think we should create a BlockModel that is only used by the controller.
The controller would then get the block and map it to a new blockModel object. This way there is no disruption to the Block object.
As for future proofing of Block vs SmartBlock serialization, we can always have custom serializers in place.

File: src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            this.blockStoreCache = blockStoreCache;
+            this.connectionManager = connectionManager;
+            this.network = network;
+            this.chain = chain;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.dateTimeProvider = dateTimeProvider;
+        }
+
+        [Route("getblock")]
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs]: `block` is enough as the endpoint name.

File: src/Stratis.Bitcoin.Features.BlockStore/Models/BlockByHashQueryModel.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public bool OutputJson { get; set; }
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.BlockStore/Models/BlockByHashQueryModel.cs]: Json should really be the default. Maybe `ShowAsHex`can be better? What do you think?

File: src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs (hunk 3 above)
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs]: Please add a comment on this method.

File: src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                return BuildErrorResponse(this.ModelState);
+            }
+
+            this.logger.LogTrace("({0}:'{1}')", nameof(ObjectByHashQueryModel.Hash), query.Hash);
+
+            try
+            {
+                var block = await this.blockStoreCache.GetBlockAsync(uint256.Parse(query.Hash)).ConfigureAwait(false);
+                if(block == null) return new NotFoundObjectResult("Block not found");
+                return query.OutputJson
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs]: I think a BlockModel here is a better option.

**Comment by monsieurleberre** [comment]
This will be done using a Model instead of changing the NBitcoin.Block code

**Pull request closed without merging**